datafusion/common/src/types/native.rs

```diff
@@ -0,0 +1,399 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use super::{
+    LogicalField, LogicalFieldRef, LogicalFields, LogicalType, LogicalUnionFields,
+    TypeSignature,
+};
+use crate::error::{Result, _internal_err};
+use arrow::compute::can_cast_types;
+use arrow_schema::{
+    DataType, Field, FieldRef, Fields, IntervalUnit, TimeUnit, UnionFields,
+};
+use std::sync::Arc;
+
+/// Representation of a type that DataFusion can handle natively. It is a subset
+/// of the physical variants in Arrow's native [`DataType`].
+#[derive(Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub enum NativeType {
```

Comment:
I was confused about the name of `NativeType` before I understood the context.  I thought the `NativeType` is a mirror of the Arrow DataType but it's not, it means the DataFusion type.
I think maybe we can just call it `Type` or `SqlType`. If I don't misunderstand it, it will be the core type system of DataFusion, right?

By the way, we have another type list in the SQL parser for the SQL API.
https://github.com/apache/datafusion-sqlparser-rs/blob/8e0d26abb3c4dda2fa091dfd9611be1dbdc14755/src/ast/data_type.rs#L36

I'm not sure if we could integrate both of them. Just mention that we need to maintain 3-level type mapping in the future. (SQL -> DataFusion -> Arrow)

Reply:
I think Type is a bit general and SqlType is confused with SqlType in parser.

It makes sense to have three kinds of types in the future.